IO_rot.f90

```diff
@@ -226,90 +222,169 @@ subroutine epilogue
+        call save_param_hdf5(Nexcite(:,:,:,5), "nexcite5.dac.", 3)
+        call save_param_hdf5(Nexcite(:,:,:,6), "nexcite6.dac.", 3)
+      endif
+      ! include viscosity results
+      if((flag_visc.ge.1).and.(vs_sav.eq.0)) then
+        call save1param(visc(:,:,:,1),tno//"viscx.dac.",1)
+        call save1param(visc(:,:,:,2),tno//"viscy.dac.",1)
+        call save1param(visc(:,:,:,3),tno//"viscz.dac.",1)
+        ! analogous save commands for HDF5 files
+        call save_param_hdf5(visc(:,:,:,1), "viscx.dac.", 3)
+        call save_param_hdf5(visc(:,:,:,2), "viscy.dac.", 3)
+        call save_param_hdf5(visc(:,:,:,3), "viscz.dac.", 3)
       endif
-       if((flag_visc.ge.1).and.(vs_sav.eq.0)) then
-          call save1param(visc(:,:,:,1),tno//"viscx.dac.",1)
-          call save1param(visc(:,:,:,2),tno//"viscy.dac.",1)
-          call save1param(visc(:,:,:,3),tno//"viscz.dac.",1)
-       endif
     endif
+
     if(flag_pip.eq.1 .or.flag_mhd.eq.0) then
-       do i=1,nvar_h
-          call save1param(U_h(:,:,:,i),tno//trim(file_h(i)),1)
-       enddo
+      do i=1,nvar_h
+        call save1param(U_h(:,:,:,i),tno//trim(file_h(i)),1)
+        call save_param_hdf5(U_h(:,:,:,i), trim(file_h(i)), 3)
+      enddo
     endif
-    if(flag_divb.eq.1 .and. flag_mhd.eq.1 .and. ps_sav .eq.0) then    
-       call save1param(U_m(:,:,:,9),tno//trim(file_m(9)),1)
+    ! Save divergence of B-field values (currently used in testing)
+    if(flag_divb.eq.1 .and. flag_mhd.eq.1 .and. ps_sav .eq.0) then
+      call save1param(U_m(:,:,:,9),tno//trim(file_m(9)),1)
+      call save_param_hdf5(U_m(:,:,:,9), trim(file_m(9)), 3)
     endif
-    
+
   end subroutine save_varfiles
 
+
   subroutine save1param(q,name,nvar)
     integer,intent(in)::nvar
     double precision, intent(in) :: q(ix,jx,kx,nvar)
     character(*), intent(in) :: name
     integer, parameter :: mf_q = 999
     integer nn
-    
+
     call dacdef3s(mf_q,trim(outdir) // '/' // name // cno,6,0)
     do nn=1,nvar
        write(mf_q) q(:,:,:,nn)
     enddo
 !print*,mf_q,name,q(1,1,1,:)
     close(mf_q)
   end subroutine save1param
-  
+
+  !
+  ! Saving a single variable array (of arbitrary rank) to an hdf5 file
+  !
+  subroutine save_param_hdf5(data, varname, rank)
+    !
+    ! This is a refactor of iot_rot.save1param, using the HDF5 library
+    ! e.g. - save_param_hdf5(visc(:,:,:,1), 'viscx', 3, (/ix,jx,kx/))
+    !
+    integer, intent(in) :: rank
+    integer(SIZE_T), dimension(rank) :: dims
+    double precision, dimension(*) :: data
+    character(*), intent(in) :: varname
+
+    integer :: error                ! Error flag
+    integer(HID_T) :: file_id       ! File identifier
+    integer(HID_T) :: dset_id       ! Dataset identifier
+    integer(HID_T) :: dspace_id     ! Dataspace identifier
+    character(len=40) :: fpath      ! Relative path for output file
+
+    ! Define file path based on rank type
+    if (rank.eq.3) then
```

Comment:
It is good to see type casting moved into the subroutine. I guess ix,iy,iz are globals. This something Ben that might know better, but my concern with this approach is whether ix,iy,iz would need to be different when collecting all the data from all the processors? When I was thinking about type casting within the subroutine, I was thinking about doing something like this: h5screate_simple_f(rank, INT(dims, KIND=8), dspace_id, error). That would keep it general, but needs testing as I am not sure it will work.

Reply:
ix, iy, iz are local variables. You might need to create a global version (ixg?) that would need to be allocated in MPI_rot.f90 (before line 26 which is where ix, iy, iz become local). There is also the set_mpi_pos subroutine that is used to determine the location of each processor so this could be useful when collating the data.